src/reanimated2/Hooks.js

```diff
@@ -304,36 +304,41 @@ export function useAnimatedGestureHandler(handlers) {
   return useEvent(
     event => {
       'worklet';
-      const UNDETERMINED = 0;
       const FAILED = 1;
+      const BEGAN = 2;
       const CANCELLED = 3;
       const ACTIVE = 4;
       const END = 5;
 
-      if (event.oldState === UNDETERMINED && handlers.onStart) {
+      if (event.state === BEGAN && handlers.onStart) {
         handlers.onStart(event, context);
       }
       if (event.state === ACTIVE && handlers.onActive) {
         handlers.onActive(event, context);
       }
-      if (event.oldState === ACTIVE && event.state === END && handlers.onEnd) {
+      if (
+        (event.oldState === BEGAN || event.oldState === ACTIVE) &&
+        event.state === END && handlers.onEnd
+      ) {
         handlers.onEnd(event, context);
       }
       if (
-        event.oldState === ACTIVE &&
-        event.state === FAILED &&
-        handlers.onFail
+        (event.oldState === BEGAN || event.oldState === ACTIVE) &&
```

Comment:
According to this https://docs.swmansion.com/react-native-gesture-handler/docs/state.html#state-flows it should check only for Began